realtek: dsa: rtl93xx: Implement MSTP state support

ecsv opened the pull request:
The MSTP support (usually implemented by mstpd) requires from the kernel that a VLAN can associated with an MSTI. At the moment, all these VLANs just get the MSTI 0 harcoded on creation. But the `vlan_tables_read()`+`vlan_tables_write()` helper already allow the modification of the MSTI and only require a minimal hook to expose this functionality.

It is also necessary to adjust the (M)STP states per MSTI and not only per port (or for the CIST). The rtl83xx_port_stp_state_set() function was in theory already capable to modify other MSTIs than CIST - if the MSTI would not have been hardcoded to 0.

The userspace can trigger these modifications using netlink:

* (Re)associating VLANs with an MSTI:

      bridge vlan global set dev <BR> vid <X> msti <Y>

* Setting the port state in a given MSTI:

      bridge mst set dev <PORT> msti <Y> state <Z>

But the HW was configured via  the `ST_CTRL` to always assume the `CIST` (MSTI 0) and ignore any other configuration for VLANs. This must then of course be disabled - which should not be a problem because the VLANs are currently always created with CIST (MSTI 0).

---

Also, the DSA port code is trying to flush associated VLANs whenever the MST  state is changed. This functionality is available on a per port+vid based using the `L2_TBL_FLUSH_CTRL` which is already used for the `.port_fast_age` callbacks.

An implementation specific for the VLAN therefore only required to toggle the VLAN bit and to write the VLAN in the request.

---

To get MSTP working at the end, it is of course also required to run a `MSTP` compatible userspace implementation (which implements `/sbin/bridge-stp`) instead of the old kernel STP implementation from the bridge code. `mstpd` should be used (and it must be checked that `/sys/class/net/switch/bridge/stp_state` is `2` for the userspace STP).

The userspace implementation also requires that the per port transmission work (#20208) and that BPDUs are trapped to the CPU (#20414)

When testing it, it is highly recommended to disable L3 offloading (#20208) to avoid weird offloading problems.

IssamHamdi committed:
realtek: dsa: rtl93xx: Switch to MSTP compatible STP mode

The realtek DSA switch driver sets up all VLANs using CIST. It is therefore
not necessary to enforce CIST using the ST_CTRL register.

This allows us later to overwrite the MSTI of VLANs. This is necessary to
get MSTP working on RTL93xx.

Signed-off-by: Issam Hamdi <ih@simonwunderlich.de>
Signed-off-by: Sven Eckelmann <se@simonwunderlich.de>
Link: https://github.com/openwrt/openwrt/pull/20421
Signed-off-by: Hauke Mehrtens <hauke@hauke-m.de>

IssamHamdi committed:
realtek: dsa: rtl93xx: Implement vlan fast age flushing

The DSA port code is trying to flush associated VLANs whenever the MST
state is changed. This functionality is available on a per port+vid based
using the L2_TBL_FLUSH_CTRL which is already used for the .port_fast_age
callbacks.

Signed-off-by: Issam Hamdi <ih@simonwunderlich.de>
Co-developed-by: Sven Eckelmann <se@simonwunderlich.de>
Signed-off-by: Sven Eckelmann <se@simonwunderlich.de>
Link: https://github.com/openwrt/openwrt/pull/20421
Signed-off-by: Hauke Mehrtens <hauke@hauke-m.de>

IssamHamdi committed:
realtek: dsa: rtl93xx: Support multi spanning tree states

The MSTP support (usually implemented by mstpd) requires from the kernel
that a VLAN can associated with an MSTI. At the moment, all these VLANs
just get the msti 0 harcoded on creation. But the
vlan_tables_read()+vlan_tables_write() helper already allow the
modification of the MSTI and only require a minimal hook to expose this
functionality.

It is also necessary to adjust the (M)STP states per MSTI and not only per
port (or for the CIST). The rtl83xx_port_stp_state_set() function was in
theory already capable to modify other MSTIs than CIST - if the msti would
not have been hardcoded to 0.

The userspace can trigger these modifications using netlink:

* (Re)associating VLANs with an MSTI:

      bridge vlan global set dev <BR> vid <X> msti <Y>

* Setting the port state in a given MSTI:

      bridge mst set dev <PORT> msti <Y> state <Z>

Signed-off-by: Issam Hamdi <ih@simonwunderlich.de>
Co-developed-by: Sven Eckelmann <se@simonwunderlich.de>
Signed-off-by: Sven Eckelmann <se@simonwunderlich.de>
Link: https://github.com/openwrt/openwrt/pull/20421
Signed-off-by: Hauke Mehrtens <hauke@hauke-m.de>

plappermaul commented:
Something like that (no favour here). Putting the locking into the outside function and we have another bit of code cleanly refactored. Thank you.

Neustradamus commented:
@ecsv: Really good job, thanks!

Djfe commented on target/linux/realtek/files-6.12/drivers/net/dsa/rtl83xx/dsa.c:
```diff
@@ -1922,6 +1934,37 @@ static int rtl83xx_vlan_del(struct dsa_switch *ds, int port,
+		return -EOPNOTSUPP;
+
+	mutex_lock(&priv->reg_mutex);
+	ret = priv->r->vlan_port_fast_age(priv, port, vid);
```
What does this do for RTL83xx type switches? I mean it recursively calls itself.
vlan_port_fast_age points to rtldsa_port_vlan_fast_age which calls vlan_port_fast_age again.
`ret` never gets set to a value.

ecsv commented on target/linux/realtek/files-6.12/drivers/net/dsa/rtl83xx/rtl931x.c:
```diff
@@ -1415,6 +1415,24 @@ static void rtl931x_vlan_port_pvid_set(int port, enum pbvlan_type type, int pvid
+	sw_w32(vid << 20, RTL931X_L2_TBL_FLUSH_CTRL + 4);
+
+	val = 0;
+	val |= port << 11;
```
> Why are port and vid switched in comparison to `rtldsa_930x_vlan_port_fast_age`?

Because of realtek: https://svanheule.net/realtek/longan/register/l2_tbl_flush_ctrl vs https://svanheule.net/realtek/mango/register/l2_tbl_flush_ctrl

> Should it be possible to unify the code with the regular fast_age functions that are not vlan related? Or does that make the code to complicated to read?

Makes it complicated to read. And we then need 3 functions - one common one and two which are the wrapper two to call this minimal common code. And most likely a configuration struct to define the things which need to be flushed and which not. I don't say that it should never be the a common helper function but this looked like overkill for me.

Djfe replied:
ah, thanks for explaining, sounds reasonable :)

KanjiMonster commented:
> Just my own notes (not a rebuttal - might also just repeat statements by the original author but with additional things to keep track of details):
> 
> > A few notes about in-kernel MSTI support I encountered while working on it for b53:
> > ```
> > * I highly recommend using a msti to internal id mapping, as MTSIDs have a 4k range, just like vlan IDs
> > ```
> 
> I had a look at b53 and mv88e6xxx and this definitely seems to a good thing to have. The current implementation is even handling the number of MSTIs wrong (see below).
> 
> > ```
> > * if a vlan has no members anymore on a bridge, its MSTI will reset to 0 / CIST, but there is no such notification to the driver (or userspace)
> > ```
> 
> Refers to the additional mapping (or at least partially needs it). If it is there:
> 
>     * moving a VLAN from an MSTI to another one requires that the old MSTI has its refcnt decreased (`vlan_msti_set`)
> 
>     * removing the last port belonging to VLAN and needs a decrease in the refcnt (`port_vlan_del`)
> 
> 
> CIST (0) must either be handled special (never reassigned/freed) or by default have reference count set to number of vlans (b53).
> 
> > ```
> > * There is no (dsa/switchdev/bridge) code currently unblocking/setting MST STP states for stand alone ports, or port joining or leaving a bridge, which means
> >   
> >   * You need to make sure that standalone ports are always forwarding in all MSTs
> >   * You need to disable/block them in all MSTs on joining a bridge
> >   * You need to unblock them when leaving a bridge
> > ```
> 
> Refers to the standard MSTI handling (not related to the mapping). In this PR, following is missing:
> 
>     * mv88e6xxx
>       
>       * when a MST is initially created, all ports need to be set to blocking (user ports), disabled (DSA port)
> 
>     * b53
>       
>       * when a MST is initially created, all ports need to be set to disabled (user ports in bridge), forwarding (user ports outside of bridge)

If one want to be really correct, this probably should be DISABLED/FORWARDING for user ports depending on whether the ports are up.

>       * Set all MSTs > 0 to DISABLED on join (`port_bridge_join`)
>       * Set all MSTs > 0 to FORWARDING on leave (`port_bridge_leave`)


And then set the same state in all MSTs when setting the (CIST) STP state on unbridged ports.

> 
> 
> Regarding the number of available MSTI on RTL devices:
> 
>     * RTL8380: 64 https://svanheule.net/realtek/maple/table/vlan
> 
>     * RTL8390: 256 https://svanheule.net/realtek/cypress/table/vlan
> 
>     * RTL930x: 64 https://svanheule.net/realtek/longan/table/vlan
> 
>     * RTL931x: 128 https://svanheule.net/realtek/longan/table/vlan
> 
> 
> ```
> sources/rtk-dms1250/src/hal/chipdef/chipdef.c:static rt_register_capacity_t rtl8390m_capacityInfo =
> sources/rtk-dms1250/src/hal/chipdef/chipdef.c:    .max_num_of_msti                = 256,
> sources/rtk-dms1250/src/hal/chipdef/chipdef.c:static rt_register_capacity_t rtl8380m_capacityInfo =
> sources/rtk-dms1250/src/hal/chipdef/chipdef.c:    .max_num_of_msti                = 64,
> sources/rtk-dms1250/src/hal/chipdef/chipdef.c:static rt_register_capacity_t rtl9310_capacityInfo =
> sources/rtk-dms1250/src/hal/chipdef/chipdef.c:    .max_num_of_msti                = 128,
> sources/rtk-dms1250/src/hal/chipdef/chipdef.c:    .max_num_of_msti                = 128,
> sources/rtk-dms1250/src/hal/chipdef/chipdef.c:static rt_register_capacity_t rtl9300_capacityInfo =
> sources/rtk-dms1250/src/hal/chipdef/chipdef.c:    .max_num_of_msti                = 64,
> ```
> 
> I am super thankful for this comment. And I am sorry for not having time to actually checking it in detail earlier.

Note that the Standard (802.1Q-2022) says in 13.14:

> No more than 64 MSTI Configuration Messages shall be encoded in an MST BPDU, and no more than 64 MSTIs shall be supported by an MST Bridge.

So anything more than 64 is purely academical from a MSTP perspective. You would only need it for PvSTP for Cisco interoperability (which at least MSTPD does not support).

Shouldn't stop the kernel from supporting more than that though.

ecsv committed:
realtek: dsa: Record number of supported MSTs

Each SoC supports a different number of MST(I)s. The code must know this
limitation to correctly reject unsupported MSTIs or to allocate a large
enough mapping table.

Signed-off-by: Sven Eckelmann <se@simonwunderlich.de>
Link: https://github.com/openwrt/openwrt/pull/20421
Signed-off-by: Hauke Mehrtens <hauke@hauke-m.de>

ecsv committed:
realtek: dsa: Add MSTI to HW MST ID mapping

The MSTI range is 0..4095 but the HW range is only supporting a lower
range - for example 0..63 for RTL930x. But the HW doesn't really need to
know the actual MSTI. It is therefore possible to use a mapping from MSTI
to HW slot to allow a larger range of MSTIs.

Since the CIST (MSTI 0) is always needed, the mapping data structure is
skipping this entry and is always keeping the HW slot 0 for CIST.

This doesn't increase the total number of MSTIs a HW supports.

Suggested-by: Jonas Gorski <jonas.gorski@gmail.com>
Signed-off-by: Sven Eckelmann <se@simonwunderlich.de>
Link: https://github.com/openwrt/openwrt/pull/20421
Signed-off-by: Hauke Mehrtens <hauke@hauke-m.de>

ecsv committed:
realtek: dsa: Sync CIST with MSTI state for unbridged ports

The VLANs and their MSTIs are shared on the realtek switch HW between
bridged and unbridged ports. But the MSTI state cannot be updated for an
unbridged port via DSA. To ensure that the port is still configured
correctly after leaving a bridge, the CIST state updates via DSA must also
be propagated to the MSTI states.

Suggested-by: Jonas Gorski <jonas.gorski@gmail.com>
Signed-off-by: Sven Eckelmann <se@simonwunderlich.de>
Link: https://github.com/openwrt/openwrt/pull/20421
Signed-off-by: Hauke Mehrtens <hauke@hauke-m.de>

ecsv committed:
realtek: dsa: Automatically return lost VLANs to CIST

If a VLAN doesn't have any members anymore, then it is removed and
implicitly returns back from any MSTI to CIST. The DSA layer will not
create any call to .vlan_msti_set and the driver is required to handle this
directly.

Suggested-by: Jonas Gorski <jonas.gorski@gmail.com>
Signed-off-by: Sven Eckelmann <se@simonwunderlich.de>
Link: https://github.com/openwrt/openwrt/pull/20421
Signed-off-by: Hauke Mehrtens <hauke@hauke-m.de>

ecsv committed:
realtek: dsa: Adjust MSTP states after joining/leaving bridge

When joining a bridge or leaving a bridge, the CIST state will
automatically be adjusted by DSA using .port_stp_state_set(). But MSTIs are
completely unhandled.

If a port is joining a bridge, the default state must be disabled. The MSTP
daemon is then responsible for adjusting the state.

If the bridge is left, the forwarding state must be enforced because VLANs
(and with this also the MSTIs assigned to them) are shared between bridged
and non-bridged ports. An unbridged port must therefore not be left in an
blocked/disabled state for a VLAN (MSTI).

Suggested-by: Jonas Gorski <jonas.gorski@gmail.com>
Signed-off-by: Sven Eckelmann <se@simonwunderlich.de>
Link: https://github.com/openwrt/openwrt/pull/20421
Signed-off-by: Hauke Mehrtens <hauke@hauke-m.de>

Djfe reviewed (approved):
Thanks for addressing my false comments. Helped me a lot :)
Looking forward to seeing this making its way into the dsa driver.
Thanks for your effort! @IssamHamdi @ecsv
I couldn't see anything else wrong with this so far while reading your changes. Approved :)
